lib/Transforms/LLZKR1CSLoweringPass.cpp

```diff
@@ -0,0 +1,688 @@
+          r1cs::FeltAttr::get(builder.getContext(), 0)
+      );
+    }
+
+    return result;
+  }
+
+  void buildAndEmitR1CS(
+      ModuleOp &moduleOp, StructDefOp &structDef, FuncDefOp &constrainFunc,
+      DenseMap<Value, unsigned> &degreeMemo
+  ) {
+    SmallVector<R1CSConstraint, 16> constraints;
+    constrainFunc.walk([&](EmitEqualityOp eqOp) {
+      OpBuilder builder(eqOp);
+      R1CSConstraint eq = lowerEquationToR1CS(eqOp.getLhs(), eqOp.getRhs(), degreeMemo);
+      constraints.push_back(eq);
+    });
+    moduleOp->setAttr("veridise.lang", StringAttr::get(moduleOp.getContext(), "r1cs"));
+    Block &entryBlock = constrainFunc.getBody().front();
+    IRMapping valueMap;
+    Location loc = structDef.getLoc();
+    OpBuilder topBuilder(moduleOp.getBodyRegion());
+
+    // Validate struct fields are felt and prepare signal types for circuit result types
+    bool hasPublicSignals = false;
+    for (auto field : structDef.getFieldDefs()) {
+      if (!field.getType().isa<FeltType>()) {
+        field.emitError("Only felt fields are supported as output signals").report();
+        signalPassFailure();
+        return;
+      }
+      if (field.isPublic()) {
+        hasPublicSignals = true;
+      }
+    }
+
+    if (!hasPublicSignals) {
+      structDef.emitError("Struct should have at least one public output").report();
+    }
+    llvm::SmallVector<mlir::NamedAttribute> argAttrPairs;
+
+    for (auto [i, arg] : llvm::enumerate(llvm::drop_begin(entryBlock.getArguments(), 1))) {
+      if (constrainFunc.hasArgPublicAttr(i + 1)) {
+        auto key = topBuilder.getStringAttr(std::to_string(i));
+        auto value = r1cs::PublicAttr::get(moduleOp.getContext());
+        argAttrPairs.emplace_back(key, value);
+      }
+    }
+    auto dictAttr = topBuilder.getDictionaryAttr(argAttrPairs);
+    auto circuit =
+        topBuilder.create<r1cs::CircuitDefOp>(loc, structDef.getSymName().str(), dictAttr);
+
+    Block *circuitBlock = circuit.addEntryBlock();
+
+    OpBuilder bodyBuilder = OpBuilder::atBlockEnd(circuitBlock);
+
+    // Step 3: Validate that all parameters to the constrain function are felt types
+    for (auto [i, arg] : llvm::enumerate(llvm::drop_begin(entryBlock.getArguments(), 1))) {
+      if (!arg.getType().isa<FeltType>()) {
+        constrainFunc.emitOpError("All input arguments must be of felt type").report();
+        signalPassFailure();
+        return;
+      }
+      auto signalType = r1cs::SignalType::get(moduleOp.getContext());
+      auto blockArg = circuitBlock->addArgument(signalType, loc);
+      valueMap.map(arg, blockArg);
+    }
+
+    // Step 4: For every struct field we a) create a signaldefop and b) add that signal to our
+    // outputs
+    DenseMap<StringRef, Value> fieldSignalMap;
+    uint32_t signalDefCntr = 0;
+    for (auto field : structDef.getFieldDefs()) {
+      r1cs::PublicAttr pubAttr;
+      if (field.hasPublicAttr()) {
+        pubAttr = r1cs::PublicAttr::get(moduleOp.getContext());
```

Comment:
nit: simplify with builder function
```suggestion
        pubAttr = bodyBuilder.getAttr<r1cs::PublicAttr>();
```